client/network/src/protocol/generic_proto/upgrade/notifications.rs

```diff
@@ -100,24 +100,25 @@ pub struct NotificationsOutSubstream<TSubstream> {
 
 impl NotificationsIn {
 	/// Builds a new potential upgrade.
-	pub fn new(protocol_name: impl Into<Cow<'static, [u8]>>) -> Self {
+	pub fn new(protocol_name: impl Into<Cow<'static, str>>) -> Self {
 		NotificationsIn {
 			protocol_name: protocol_name.into(),
 		}
 	}
 
 	/// Returns the name of the protocol that we accept.
-	pub fn protocol_name(&self) -> &[u8] {
+	pub fn protocol_name(&self) -> &Cow<'static, str> {
 		&self.protocol_name
 	}
 }
 
 impl UpgradeInfo for NotificationsIn {
-	type Info = Cow<'static, [u8]>;
+	type Info = Vec<u8>;
 	type InfoIter = iter::Once<Self::Info>;
 
 	fn protocol_info(&self) -> Self::InfoIter {
-		iter::once(self.protocol_name.clone())
+		// TODO: Can we do better?
+		iter::once(self.protocol_name.as_bytes().to_vec())
```

Comment:
I am not sure how to handle this without an allocation. In theory the following should work, but the compiler has difficulties inferring the static lifetime.

```suggestion
		iter::once(Cow::Borrowed(self.protocol_name.as_bytes()))
```

Ideas?

Reply:
Even though the data inside the Cow is bounded by `'static`, the borrow you get from deref will be bounded by the lifetime of `self`. I think `self` would have to be `'static` for this to work. I don't have any better suggestion than allocating.

Reply:
> Even though the data inside the Cow is bounded by 'static, the borrow you get from deref will be bounded by the lifetime of self.

Thanks @andresilva. That helped. The below compiles and prevents the allocation in the standard (borrowed) case.

```diff
diff --git a/client/network/src/protocol/generic_proto/upgrade/notifications.rs b/client/network/src/protocol/generic_proto/upgrade/notifications.rs
index b245e71af..32dae1afd 100644
--- a/client/network/src/protocol/generic_proto/upgrade/notifications.rs
+++ b/client/network/src/protocol/generic_proto/upgrade/notifications.rs
@@ -259,12 +259,15 @@ impl NotificationsOut {
 }
 
 impl UpgradeInfo for NotificationsOut {
-       type Info = Vec<u8>;
+       type Info = Cow<'static, [u8]>;
        type InfoIter = iter::Once<Self::Info>;
 
        fn protocol_info(&self) -> Self::InfoIter {
-               // TODO: Can we do better?
-               iter::once(self.protocol_name.as_bytes().to_vec())
+               let bytes: Cow<'static, [u8]> = match &self.protocol_name {
+                       Cow::Borrowed(s) => Cow::Borrowed(s.as_bytes()),
+                       Cow::Owned(s) => Cow::Owned(s.as_bytes().to_vec())
+               };
+               iter::once(bytes)
        }
 }
```

Reply:
Nice, destructuring the `Cow` like that allows us to get the original lifetime of the borrowed data, didn't think of that 👍